network/p2p/dht_test.go

```diff
@@ -0,0 +1,219 @@
+	// hence expect count and not count - 1 messages to be received (one by each node, including the sender)
+	ch := make(chan flow.Identifier, count)
+	for _, n := range nodes {
+		m := n.id
+		// defines a func to read from the subscription
+		subReader := func(s *pubsub.Subscription) {
+			msg, err := s.Next(suite.ctx)
+			require.NoError(suite.T(), err)
+			require.NotNil(suite.T(), msg)
+			assert.Equal(suite.T(), []byte("hello"), msg.Data)
+			ch <- m
+		}
+
+		// Subscribes to the test topic
+		s, err := n.Subscribe(suite.ctx, topic)
+		require.NoError(suite.T(), err)
+
+		// kick off the reader
+		go subReader(s)
+
+	}
+
+	// fullyConnectedGraph checks that each node is directly connected to all the other nodes
+	fullyConnectedGraph := func() bool {
+		for i := 0; i < len(nodes); i++ {
+			for j := i + 1; j < len(nodes); j++ {
+				if nodes[i].host.Network().Connectedness(nodes[j].host.ID()) == network.NotConnected {
+					return false
+				}
+			}
+		}
+		return true
+	}
+	// assert that the graph is fully connected
+	require.Eventually(suite.T(), fullyConnectedGraph, time.Second*5, tickForAssertEventually, "nodes failed to discover each other")
+
+	// Step 4: publish a message to the topic
+	require.NoError(suite.T(), dhtServerNode.Publish(suite.ctx, topic, []byte("hello")))
+
+	// Step 5: By now, all peers would have been discovered and the message should have been successfully published
+	// A hash set to keep track of the nodes who received the message
+	recv := make(map[flow.Identifier]bool, count)
+	for i := 0; i < count; i++ {
+		select {
+		case res := <-ch:
+			recv[res] = true
+		case <-time.After(3 * time.Second):
+			var missing flow.IdentifierList
+			for _, n := range dhtClientNodes {
+				if _, found := recv[n.id]; !found {
+					missing = append(missing, n.id)
+				}
+			}
+			assert.Fail(suite.T(), " messages not received by nodes: "+strings.Join(missing.Strings(), ","))
+			break
+		}
+	}
+
+	// Step 6: unsubscribes all nodes from the topic
+	for _, n := range nodes {
+		assert.NoError(suite.T(), n.UnSubscribe(topic))
+	}
+}
+
+// CreateNode creates the given number of libp2pnodes
+// if dhtServer is true, the DHTServer is used as for Discovery else DHTClient
+func (suite *DHTTestSuite) CreateNodes(count int, dhtServer bool) (nodes []*Node) {
+
+	// keeps track of errors on creating a node
+	var err error
+	logger := log.Output(zerolog.ConsoleWriter{Out: os.Stderr}).With().Caller().Logger()
+	defer func() {
+		if err != nil && nodes != nil {
+			// stops all nodes upon an error in starting even one single node
+			suite.StopNodes(nodes)
+		}
+	}()
+
+	handlerFunc := func(network.Stream) {}
+
+	// creating nodes
+	for i := 1; i <= count; i++ {
+		_, key := generateNetworkingAndLibP2PKeys(suite.T())
+		noopMetrics := metrics.NewNoopCollector()
+
+		pingInfoProvider, _, _ := MockPingInfoProvider()
+
+		connManager := NewConnManager(logger, noopMetrics)
+		libP2PHost, err := LibP2PHost(context.Background(), "0.0.0.0:0", key,
```

Comment:
Use `suite.ctx` here? it's a unittest so I'm not so concerned about antipatterns here